src/utils/tree_node.C

```diff
@@ -485,40 +493,33 @@ const Elem* TreeNode<N>::find_element(const Point& p, const std::set<subdomain_i
 
 
 template <unsigned int N>
-const Elem* TreeNode<N>::find_element_in_children(const Point& p, const std::set<subdomain_id_type> *allowed_subdomains) const
+const Elem* TreeNode<N>::find_element_in_children
+(const Point& p,
+ const std::set<subdomain_id_type> *allowed_subdomains,
+ Real relative_tol) const
 {
   libmesh_assert (!this->active());
 
-  unsigned int excluded_child = libMesh::invalid_uint;
+  std::vector<bool> searched_child(children.size(), false);
```

Comment:
The C++ standard specifies otherwise.  Concurrent const operations on C++ containers are required to be thread-safe.

http://stackoverflow.com/questions/10833512/safe-parallel-read-only-access-to-a-stl-container?rq=1

The exception for vector<bool> only refers to non-const modifications.

Do you have any more details on what you encountered in MOOSE?  It wouldn't be the first time we'd had to work around a common compiler bug.

Or for that matter it might not be a bug, technically.  All the above requirements are from the C++11 standard.  The C++98 standard specifies "What are threads?  You're on your own."

The reason for wanting to deprecate vector<bool> (and replace it with some non-STL-looking bitset class, not with "ha-ha, you have to use 8 times more RAM now!") dates back to the weird proxy references it needs, not multithreading concerns.  Right now you can write something as simple as `T& = v[0]` and it will compile for every T *except* bool.